extensions/vi-mode/tests/utils.lisp

```diff
@@ -0,0 +1,298 @@
+                               :start (1- (if (< buffer-pos (position-at-point start))
+                                              (+ (position-at-point start) 2)
+                                              (position-at-point start)))
+                               :end (1- (if (< buffer-pos (position-at-point end))
+                                            (+ (position-at-point end) 2)
+                                            (position-at-point end))))
+                 (write-char #\> s)
+                 (setf read-pos
+                       (if (< buffer-pos (position-at-point end))
+                           (+ (position-at-point end) 2)
+                           (position-at-point end)))))))
+          buf-str))))
+
+(defun make-buffer-string (buffer)
+  (%make-buffer-string (buffer-text buffer)
+                       (position-at-point (buffer-point buffer))))
+
+(defun text-backslashed (text)
+  (ppcre:regex-replace-all "[\\n\\r\\t]" text
+                           (lambda (matches)
+                             (ecase (aref matches 0)
+                               (#\Newline "\\n")
+                               (#\Return "\\r")
+                               (#\Tab "\\t")))
+                           :simple-calls t))
+
+(defun parse-command-keys (keys-string)
+  (check-type keys-string string)
+  (let (keys)
+    (ppcre:do-matches-as-strings (key-str "(?<!\\\\)<[^\\>]+?>|." keys-string)
+      (push
+       (if (= (length key-str) 1)
+           (make-key :sym key-str)
+           (let (key-args)
+             (ppcre:do-register-groups (modifier)
+                 ("(H|S|M|C|Shift)-" key-str)
+               (cond
+                 ((string= modifier "H")
+                  (appendf key-args '(:hyper t)))
+                 ((string= modifier "S")
+                  (appendf key-args '(:super t)))
+                 ((string= modifier "M")
+                  (appendf key-args '(:meta t)))
+                 ((string= modifier "C")
+                  (appendf key-args '(:ctrl t)))
+                 ((string= modifier "Shift")
+                  (appendf key-args '(:shift t)))))
+             (let ((sym-str (ppcre:scan-to-strings "[^<-]+(?=>$)" key-str)))
+               (apply #'make-key :sym (if-let (char (name-char sym-str))
+                                        (string char)
+                                        sym-str)
+                      key-args))))
+       keys))
+    (nreverse keys)))
+
+(defmacro with-test-buffer ((var buffer-string
+                             &rest buffer-args
+                             &key name (temporary t temporary-specified-p)
+                             &allow-other-keys)
+                            &body body)
+  (declare (ignore temporary))
+  (unless temporary-specified-p
+    (setf (getf buffer-args :temporary) t))
+  (remove-from-plistf buffer-args :name)
+  (with-gensyms (point buffer-content position visual-regions)
+    `(with-fake-interface ()
```

Comment:
I am a little concerned about the with-fake-interface being hidden inside this macro.
The person reading the test needs to dig into the macro to understand that this is using a fake-interface.
How about moving the use of with-fake-interface to the top level of testing?